lib/internal/url.js

```diff
@@ -932,6 +935,64 @@ function urlToOptions(url) {
   return options;
 }
 
+function getPathFromURLWin32(url) {
+  var hostname = url.hostname;
+  var pathname = url.pathname;
+  for (var n = 0; n < pathname.length; n++) {
+    if (pathname[n] === '%') {
+      var third = pathname.codePointAt(n + 2) | 0x20;
+      if ((pathname[n + 1] === '2' && third === 102) || // 2f 2F /
+          (pathname[n + 1] === '5' && third === 99)) {  // 5c 5C \
+        return new Error('Path must not include encoded \\ or / characters');
+      }
+    }
+  }
+  pathname = decodeURIComponent(pathname);
+  if (hostname !== '') {
+    // If hostname is set, then we have a UNC path
+    // Pass the hostname through domainToUnicode just in case
+    // it has been punycode encoded. We do not need to worry
+    // about percent encoding because the URL parser will have
+    // already taken care of that for us.
+    return `//${domainToUnicode(hostname)}${pathname}`;
+  } else {
+    // Otherwise, it's a local path that requires a drive letter
+    var letter = pathname.codePointAt(1) | 0x20;
+    var sep = pathname[2];
+    if (letter < 97 || letter > 122 ||   // a..z A..Z
+        (sep !== ':' && sep !== '|')) {
+      return new Error('File URLs must specify absolute paths');
+    }
+    return pathname.slice(1);
+  }
+}
+
+function getPathFromURLPosix(url) {
+  if (url.hostname !== '') {
+    return new Error(
+      `File URLs on ${os.platform()} must use hostname localhost`);
```

Comment:
Yep. For file URLs, the URL parser normalizes 'localhost' to the empty string ''. They are equivalent.

```js
> new url.URL('file://localhost/a/b/c');                                                                                                                                                                                                               1
URL {                                                                                                                                                                                                                                                  1
  href: file:///a/b/c
  protocol: file:
  pathname: /a/b/c
}
> new url.URL('file:///a/b/c');
URL {
  href: file:///a/b/c
  protocol: file:
  pathname: /a/b/c
}
```